lib/crypt-argon2.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (C) 2020 Samanta Navarro <ferivoz@riseup.net>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "crypt-port.h"
+#include "crypt-hashes.h"
+
+#include <errno.h>
+
+#if INCLUDE_argon2_d || INCLUDE_argon2_i || INCLUDE_argon2_id
+
+#include <argon2.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "alg-argon2-encoding.h"
+
+static uint32_t
+get_value(const char *nptr, const char *arg, char **endptr, const char end)
+{
+  unsigned long val;
+
+  if (strncmp (nptr, arg, strlen (arg)))
+    return 0;
+  nptr += strlen (arg);
+  if (!(*nptr >= '1' && *nptr <= '9'))
+    return 0;
+  val = strtoul (nptr, endptr, 10);
+  if (val == 0 || val > UINT32_MAX || **endptr != end)
+    return 0;
+  return (uint32_t) val;
+}
+
+#define GET_VALUE(dst, arg, end) \
+{ \
+        dst = get_value (s, arg, &q, end); \
+        if (dst == 0) \
+          { \
+            errno = EINVAL; \
+            return; \
+          } \
+        s = q + 1; \
+}
+
+static void
+crypt_argon2_rn (const argon2_type hash_type,
+                 const char *phrase, size_t phr_size,
+                 const char *setting, size_t set_size,
+                 uint8_t *output, size_t o_size,
+                 void *scratch, size_t s_size)
+{
+  char *buf = scratch, *q;
+  const char *d, *s = setting, *type;
+  uint8_t *salt;
+  size_t pos, salt_len;
+  uint32_t v, m, t, p;
+
+  if (s_size == 0 || s_size < phr_size || set_size < 26)
+    {
+      errno = EINVAL;
+      return;
+    }
+
+  switch (hash_type) {
+  case Argon2_d:
+    type = "$argon2d$";
+    break;
+  case Argon2_i:
+    type = "$argon2i$";
+    break;
+  case Argon2_id:
+    type = "$argon2id$";
+    break;
+  default:
+    type = NULL;
+    break;
+  }
+
+  if (type == NULL)
+    {
+      errno = EINVAL;
+      return;
+    }
+
+  /* Setting for a different hash algorithm. */
+  if (strncmp (s, type, strlen(type)))
```

Comment:
Codacy found an issue: [Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected) (CWE-126).  ](https://app.codacy.com/manual/besser82/libxcrypt/pullRequest?prid=6317411)